keps/sig-storage/3762-persistent-volume-last-phase-transition-time/README.md

```diff
@@ -0,0 +1,906 @@
+
+In general we try to use the same stages (alpha, beta, GA), regardless of how the
+functionality is accessed.
+
+[feature gate]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial e2e tests completed and enabled
+
+#### Beta
+
+- Gather feedback from developers and surveys
+- Complete features A, B, C
+- Additional tests are in Testgrid and linked in KEP
+
+#### GA
+
+- N examples of real-world usage
+- N installs
+- More rigorous forms of testing—e.g., downgrade tests and scalability tests
+- Allowing time for feedback
+
+**Note:** Generally we also wait at least two releases between beta and
+GA/stable, because there's no opportunity for user feedback, or even bug reports,
+in back-to-back releases.
+
+**For non-optional features moving to GA, the graduation criteria must include
+[conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+
+#### Deprecation
+
+- Announce deprecation and support policy of the existing flag
+- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
+- Address feedback on usage/changed behavior, provided on GitHub issues
+- Deprecate the flag
+-->
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Unit tests completed and enabled
+- Initial e2e tests completed and enabled
+
+#### Beta
+
+- Allowing time for feedback (at least 2 releases between beta and GA).
+- Add unit tests covering feature enablement/disablement.
+- Manually test version skew between the API server and KCM. See the expected
+  behavior below in Version Skew Strategy.
+- Manually test upgrade->downgrade->upgrade path.
+
+#### GA
+
+- No users complaining about the new behavior.
+
+### Upgrade / Downgrade Strategy
+
+<!--
+If applicable, how will the component be upgraded and downgraded? Make sure
+this is in the test plan.
+
+Consider the following in developing an upgrade/downgrade strategy for this
+enhancement:
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to maintain previous behavior?
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to make use of the enhancement?
+-->
+
+No change in cluster upgrade / downgrade process.
+
+When downgrading from a version that added the new timestamp field PVs we need to make sure that after downgrade the
+values of the disabled field are removed.
```

Comment:
How exactly you're going to achieve it?

I'm assuming you're talking about "eventual removal" here, really, right?
[the controller will unset it on the next processing of that object if it realized FG is disabled and field is set]

Reply:
Yes, but the term "next processing" is still quite broad, it would help to specify it more. The field presence check and resetting could be done:
1) While updating volume status (`updateVolumePhase()`) - the downside is that disabling FG will have no effect util volume status changes phase again. This does not seem optimal.
2) At any PV update - some PRs (e.g. [this one](https://github.com/kubernetes/kubernetes/pull/105963/files#diff-4e64d28e1697fcabd36d8a1d8f49dd87b23cb6d23150b3202465c4f864078482R98)) implement this by resetting value of the new field to `nil` during validation. This approach seems ok, is this a viable option for this enhancement as well? Also this presents a similar caveat as when the FG is enabled - the effect won't be immediate.
3) Make PV controller actively (on each sync) search for PVs with the new field and reset it if FG is disabled. This seems like an overkill. If the value is correct and reflects real last phase change there's no risk of persisting it in the PV objects even if FG is disabled.

Reply:
I agree that option (3) is probably an overkill - this is what I meant above as "eventual removal".

I'm not sure if (1) isn't good enough - but no matter what decision you will make, it should be explicitly mentioned in the KEP.

Also - @msau42 - for her thoughts on it.

Reply:
When a feature is disabled, the api server strategy implementation should drop disabled fields on read/write, so approach 2) should cover it.

Reply:
Thank you @wojtek-t and @msau42 for the feedback - KEP has been updated.